src/reflection.jl

```diff
@@ -17,25 +23,58 @@ function ka_code_typed(kernel, argtypes; ndrange=nothing, workgroupsize=nothing,
 end
 
 
+"""
+Get the llvm code for a kernel
+
+# Examples
+```
+@ka_code_typed kernel(args)
+@ka_code_typed optimize=false kernel(args)
+```
+Works for CPU or CUDA kernels, with static or dynamic declarations
+"""
 macro ka_code_typed(ex0...)
     ex = ()
+    args = gensym(:args)
+    old_args = nothing
+    kern = nothing
     for i = 1:length(ex0)
         if ex0[i].head == :call
+            # inside kernel() expr
             while length(ex0[i].args) > 2
-                kw = ex0[i].args[end]
-                @assert kw.head == :kw
-                kw.args[2] = esc(kw.args[2])
-                kw.head = Symbol("=")
-                resize!(ex0[i].args, length(ex0[i].args) - 1)
-                ex = (kw,)..., ex...
+                if isa(ex0[i].args[end], Expr)
+                    # at expr (like ndrange=10)
+                    kw = ex0[i].args[end]
+                    @assert kw.head == :kw
+                    kw.args[2] = esc(kw.args[2])
+                    kw.head = Symbol("=")
+                    resize!(ex0[i].args, length(ex0[i].args) - 1)
+                    ex = (kw,)..., ex...
+                else
+                    # only symbols left
+                    break
+                end
             end
+            # save kernel args
+            old_args = Expr(:tuple, map(esc, ex0[i].args[2:end])...)
+            resize!(ex0[i].args, 2)
+            ex0[i].args[2] = Expr(:..., args)
+            kern = esc(ex0[i].args[1])
         end
         ex = ex..., ex0[i]
     end
+    @assert(old_args != nothing)
+    @assert(kern != nothing)
 
     thecall = InteractiveUtils.gen_call_with_extracted_types_and_kwargs(__module__, :ka_code_typed, ex)
 
     quote
+        local $(esc(args)) = $(old_args)
+        if isa($kern, Kernel{CUDADevice})
+            # translate CuArray to CuDeviceArray
+            local $(esc(args)) = map(CUDA.cudaconvert, $(esc(args)))
```

Comment:
```suggestion
            $(esc(args)) = map(CUDA.cudaconvert, $(esc(args)))
```
I don't think you need the `local` here